apps/vscode/command_client.py

```diff
@@ -0,0 +1,315 @@
+        return {
+            "commandId": self.command_id,
+            "args": self.args,
+            "waitForFinish": self.wait_for_finish,
+            "returnCommandOutput": self.return_command_output,
+            "uuid": self.uuid,
+        }
+
+
+def write_request(request: Request, path: Path):
+    """Converts the given request to json and writes it to the file, failing if
+    the file already exists unless it is stale in which case it replaces it
+
+    Args:
+        request (Request): The request to serialize
+        path (Path): The path to write to
+
+    Raises:
+        Exception: If another process has an active request file
+    """
+    try:
+        write_json_exclusive(path, request.to_dict())
+        request_file_exists = False
+    except FileExistsError:
+        request_file_exists = True
+
+    if request_file_exists:
+        handle_existing_request_file(path)
+        write_json_exclusive(path, request.to_dict())
+
+
+def handle_existing_request_file(path):
+    stats = path.stat()
+
+    modified_time_ms = stats.st_mtime_ns / 1e6
+    current_time_ms = time.time() * 1e3
+
+    if abs(modified_time_ms - current_time_ms) < STALE_TIMEOUT_MS:
+        raise Exception("Another process has an active request file")
+    else:
+        print("Removing stale request file")
+        path.unlink()
+
+
+def run_vscode_command(
+    command_id: str,
+    *args: str,
+    wait_for_finish: bool = False,
+    return_command_output: bool = False,
+):
+    """Runs a VSCode command, using command server if available
+
+    Args:
+        command_id (str): The ID of the VSCode command to run
+        wait_for_finish (bool, optional): Whether to wait for the command to finish before returning. Defaults to False.
+        return_command_output (bool, optional): Whether to return the output of the command. Defaults to False.
+
+    Raises:
+        Exception: If there is an issue with the file-based communication, or
+        VSCode raises an exception
+
+    Returns:
+        Object: The response from the command, if requested.
+    """
+    # NB: This is a hack to work around the fact that talon doesn't support
+    # variable argument lists
+    args = [x for x in args if x is not NotSet]
+
+    username = getpass.getuser()
+
+    communication_dir_path = Path(gettempdir()) / f"vscode-command-server-{username}"
+
+    if not communication_dir_path.exists():
+        if args or return_command_output:
+            raise Exception("Must use command-server extension for advanced commands")
+        print("Communication dir not found; falling back to command palette")
+        run_vscode_command_by_command_palette(command_id)
+        return
+
+    request_path = communication_dir_path / "request.json"
+    response_path = communication_dir_path / "response.json"
+
+    # Generate uuid that will be mirrored back to us by command server for
+    # sanity checking
+    uuid = str(uuid4())
+
+    unlink_if_exists(response_path)
```

Comment:
@lunixbochs any objection to this one?  Technically it should have been removed after last request, so something is wrong if it exists here.  But there is something nice about ensuring a clean slate here

Reply:
Yep, I object to this one. The state machine is pretty well designed and this would hide bugs.